pages/a/orchestrating-crate-with-docker-112.html

```diff
@@ -0,0 +1,148 @@
+```
+
+define the endpoint mode to be 'dnsrr' (DNS round robin - the dns entry 'crate' should cycle through the available IP addresses), `update-parallelism` and `update-delay` set how many instances should be updated in parallel and with what delay. Finally there is some volume mapping to ensure Crate is not writing data into the container (*don't use /tmp 's tmpfs for storing data. It will be gone after restarting*).
+
+```bash
+  crate:latest \
+    crate \
+    -Des.discovery.zen.ping.multicast.enabled=false \
+    -Des.discovery.zen.ping.unicast.hosts=crate \
+    -Des.gateway.expected_nodes=3 \
+    -Des.discovery.zen.minimum_master_nodes=2 \
+    -Des.gateway.recover_after_nodes=2 \
+    -Des.network.bind=_eth0:ipv4_
+```
+This part is the same for many Docker deployment of Crate: Multicast is disabled (it's not supported on the overlay network), instead the Docker DNS record 'crate' (the service name) is responsible for the cluster to find the other nodes. Since we are working with a 3-node cluster, the `expected_nodes`, `minimum_master_nodes`, and `recover_after_nodes` are set accordingly. The last line tells Crate to explicitly listen and bind to the IPv4 address of eth0 (usually the crate-network NIC), which is necessary since the auto discovery on the virtual adapters might be wrong.
+
+## Next Steps
+
+This setup gives you a working cluster but by setting the endpoint-mode to `DNSRR`, Docker disables its included load balancer and does not assign a public IP or port mapping to the cluster. The _other_ setting `VIP` (virtual IP) would provide a load-balanced public endpoint, but prohibits communication between the nodes. While we (and others) consider this a bug, a fix by Docker has yet (as of mid-September 2016) to be released.
+
+However since the ports are mapped automatically, there are steps to remedy this:
+ - Use NGINX (or any reverse proxy server) to distribute requests among the swarm nodes by enumerating all container IPs in an [upstream section](http://nginx.org/en/docs/http/ngx_http_upstream_module.html)
+ - Use a firewall ([firewalld](https://docs.fedoraproject.org/en-US/Fedora/19/html/Security_Guide/sec-Configure_Port_Forwarding-CLI.html), [IPTables](https://www.digitalocean.com/community/tutorials/how-to-forward-ports-through-a-linux-gateway-with-iptables), [UFW](https://help.ubuntu.com/lts/serverguide/firewall.html) ...) to forward requests to the ports to the containers (but be aware that the overlay network also requires [certain ports](http://stackoverflow.com/questions/37312055/can-firewalld-work-with-the-docker-overlay-network))
+
+Docker's 1.12 release marks a big step forward in unifying orchestration and management tools for a containerized environment, which is also a great place to run Crate and your production microservices in. Furthermore the rolling upgrade features of the Docker swarm makes it really easy to keep Crate at the latest version without downtime or risk of data loss.
```

Comment:
remove the word "in" at the end of this sentence: "...your production microservices."